### ISSUE TITLE
GTX 1070 - No GPU found suitable for NVEnc Encoding

### ISSUE THREAD

**Issue by bumpaneer** [rigaya/NVEnc#141]
I am attempting to use version 4.45 with StaxRip and I'm running into issues with my card after upgrading to the new NVEnc version. Graphics card driver has been updated to 436.02 on Windows 10.

```
D:\DVD Tools\StaxRip\Apps\Encoders\NVEnc>NVEncC64.exe --check-hw
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Loaded nvEncodeAPI64.dll.
InitCuda: device #0.

Checking Environment Info...
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18362)
CPU            Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
No GPU found suitable for NVEnc Encoding.
Failed to initialize CUDA.
No NVEnc support.


D:\DVD Tools\StaxRip\Apps\Encoders\NVEnc>NVEncC64.exe --check-environment
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Environment Info
OS : Windows 10 x64 (18362)
CPU: Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
RAM: Used 5548 MB, Total 16341 MB
GPU: GeForce GTX 1070 (15 EU) @ 1835 MHz (436.02)
```

**Comment by rigaya** [comment]
I've updated to driver 436.02 for test but having no trouble...

x64\NVEncC64.exe --check-hw
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Loaded nvEncodeAPI64.dll.
InitCuda: device #0.

Checking Environment Info...
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18362)
CPU            Intel Core i9-7980XE @ 2.60GHz [TB: 4.10GHz] (18C/36T)
GPU            #0: GeForce RTX 2070 (436.2)
NVENC / CUDA   NVENC API 9.0, CUDA 10.1, schedule mode: unknown
cuInit: Success.
cuDeviceGetCount: Success.
cuDeviceGet: ID:0.
cuDeviceComputeCapability: Success: 7.5.
NVENC capabilities: OK.
using cuda schedule mode: auto.
cuCtxCreate: Success.
cuCtxPopCurrent: Success.
cuvidInit: Success.
cuvidCtxLockCreate: Success.
InitCuda: Success.
nvEncodeAPICreateInstance(APIVer=0x70020009): Success.
NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): Success.
nvEncDestroyEncoder: success.
Closing perf monitor...
Avaliable Codec(s)
H.264/AVC
H.265/HEVC

x64\NVEncC64.exe --check-environment
NVEncC (x64) 4.45 (r1169) by rigaya, Aug 19 2019 11:47:02 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Environment Info
OS : Windows 10 x64 (18362)
CPU: Intel Core i9-7980XE @ 2.60GHz [TB: 4.10GHz] (18C/36T)
RAM: Used 7149 MB, Total 16077 MB
GPU: #0: GeForce RTX 2070 (2304 cores, 1710 MHz)[PCIe3x16][436.2]


There might be something wrong with CUDA on your environment, but I cannot tell exactly what is the reason.

**Comment by rigaya** [comment]
The compilation of 32bit and 64bit differs, so that might be causing the difference, but it's hard to say.
- 32bit: VS2015 + CUDA 8.0
- 64bit VC2017 + CUDA 10.1

I've uploaded [NVEnc 4.46](https://github.com/rigaya/NVEnc/releases/tag/4.46), adding more log on --check-hw, which might be a clue. Would you please send me the log of it?

**Comment by rigaya** [comment]
Thank you, it's a big hint there. I'll try fixing this problem once I have time.

**Comment by rigaya** [comment]
I have [updated the app](https://github.com/rigaya/NVEnc/releases/tag/4.47) to work around the error returned by cudaGetDeviceProperties. Would you have a try?

**Comment by bumpaneer** [comment]
Looks like a different set of errors now, but it still fails when I attempt to encode.

```
D:\DVD Tools\StaxRip\NVEncC_4.47_x64>NVEncC64.exe --check-hw
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  1 2019 06:07:40 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Loaded nvEncodeAPI64.dll.
InitCuda: device #0.

Checking Environment Info...
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  1 2019 06:07:40 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18362)
CPU            Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
gpuinfo: found nvcuda.dll.
gpuinfo: cuInit(0): success.
gpuinfo: cuvidInit(0): success.
gpuinfo: Found 1 CUDA device(s).
gpuinfo: checking for device #0.
gpuinfo:   cuDeviceGet: success
gpuinfo:   cudaGetDeviceProperties: CUDA 1397414229.32759
gpuinfo:   PCIBusId:
gpuinfo:   NV Driver version: 436020.
gpuinfo:   CUDA Driver version: 10010.
gpuinfo:   getHWDecCodecCsp
GPU            #0:
▓ÅbH (436.2)
NVENC / CUDA   NVENC API 9.0, CUDA 10.1, schedule mode: auto
cuInit: Success.
cuDeviceGetCount: Success.
cuDeviceGet: ID:0.
cuDeviceComputeCapability: Success: 6.1.
NVENC capabilities: OK.
using cuda schedule mode: auto.
cuCtxCreate: Success.
cuCtxPopCurrent: Success.
cuvidInit: Success.
cuvidCtxLockCreate: Success.
InitCuda: Success.
nvEncodeAPICreateInstance(APIVer=0x70020009): Success.
Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
No NVEnc support.

D:\DVD Tools\StaxRip\NVEncC_4.47_x64>NVEncC64.exe --check-environment
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  1 2019 06:07:40 (VC 1916/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
Environment Info
OS : Windows 10 x64 (18362)
CPU: Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
RAM: Used 5136 MB, Total 16341 MB
GPU: #0: [436.2]


D:\DVD Tools\StaxRip\NVEncC_4.47_x64>NVEncC64.exe --check-device
No GPU found suitable for NVEnc Encoding.
```


> Error Video encoding using NVEnc 4.47 (2.0.3.0)
> 
> Video encoding using NVEnc 4.47 failed with exit code: 1 (0x1)
> 
> The exit code might be a system error code: STATUS_WAIT_1
> 
> The exit code might be a system error code: Incorrect function.
> 
> 
> ------------------- Video encoding using NVEnc 4.47 -------------------
> 
> "D:\DVD Tools\StaxRip\NVEncC_4.47_x64\NVEncC64.exe" --vbrhq 3000 --codec h265 --preset quality --bframes 0 --lookahead 32 -i "D:\input.mkv" -o "D:\output.h265"
> 
> gpuinfo:   Error: device does not satisfy required CUDA version (>=3.0): 686455101.32759
> No GPU found suitable for NVEnc Encoding.
> Failed to initialize CUDA.
> GPU #0 (èß¯) does not support H.265/HEVC encoding.
> Unknown erro occurred during checking GPU.
> 
> 
> 
> StaxRip.ErrorAbortException: Video encoding using NVEnc 4.47 failed with exit code: 1 (0x1)
> 
> The exit code might be a system error code: STATUS_WAIT_1
> 
> The exit code might be a system error code: Incorrect function.
> 
> 
> ------------------- Video encoding using NVEnc 4.47 -------------------
> 
> "D:\DVD Tools\StaxRip\NVEncC_4.47_x64\NVEncC64.exe" --vbrhq 3000 --codec h265 --preset quality --bframes 0 --lookahead 32 -i "D:\input.mkv" -o "D:\output.h265"
> 
> gpuinfo:   Error: device does not satisfy required CUDA version (>=3.0): 686455101.32759
> No GPU found suitable for NVEnc Encoding.
> Failed to initialize CUDA.
> GPU #0 (èß¯) does not support H.265/HEVC encoding.
> Unknown erro occurred during checking GPU.
>

**Comment by rigaya** [comment]
Actually, there is nothing I can do if nvEncOpenEncodeSessionEx() returns error, as it is required to start NVENC encoding. I've tested in 2 of the Win10 system with Geforce I have, but succeeded without error.

As you have commented that x86 version runs fine, I've thought that it might be a problem of the build. I've attached a build by VS2019 + CUDA 10.1 Update 2, the latest build environment now. (Current ordinary release build is VS2017 + CUDA 10.1 (Unmarked))
[NVEncC64_4.47_test_VS2019_CUDA10_1_Update2.zip](https://github.com/rigaya/NVEnc/files/3566266/NVEncC64_4.47_test_VS2019_CUDA10_1_Update2.zip)

**Comment by bumpaneer** [comment]
It does not throw an error on CUDA initialization anymore in StaxRip, but still errors out. One thing that is interesting is it does not return any information about the card like the 32-bit version does (card name, core count, processor speed). 

Version 4.32 had the same nvEncOpenEncodeSessionEx() error for me, but encoding still worked out of StaxRip. I am not sure what else to try, as it should only be the Nvidia drivers that need to be installed, correct?

Thanks for all the time you've spent trying to get this resolved.

```
D:\DVD Tools\StaxRip\NVEncC64_4.47_test_VS2019_CUDA10_1_Update2>NVEncC64.exe --check-hw
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  2 2019 22:03:53 (VC 1922/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Loaded nvEncodeAPI64.dll.
InitCuda: device #0.

Checking Environment Info...
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  2 2019 22:03:53 (VC 1922/Win/avx2)
OS Version     Windows 10 x64 (18362)
CPU            Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
gpuinfo: found nvcuda.dll.
gpuinfo: cuInit(0): success.
gpuinfo: cuvidInit(0): success.
gpuinfo: Found 1 CUDA device(s).
gpuinfo: checking for device #0.
gpuinfo:   cuDeviceGet: success
gpuinfo:   PCIBusId:
gpuinfo:   NV Driver version: 436020.
gpuinfo:   CUDA Driver version: 10010.
gpuinfo:   getHWDecCodecCsp
GPU            #0:  (436.2)
NVENC / CUDA   NVENC API 9.0, CUDA 10.1, schedule mode: auto
cuInit: Success.
cuDeviceGetCount: Success.
cuDeviceGet: ID:0.
cuDeviceComputeCapability: Success: 6.1.
NVENC capabilities: OK.
using cuda schedule mode: auto.
cuCtxCreate: Success.
cuCtxPopCurrent: Success.
cuvidInit: Success.
cuvidCtxLockCreate: Success.
InitCuda: Success.
nvEncodeAPICreateInstance(APIVer=0x70020009): Success.
Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
No NVEnc support.

D:\DVD Tools\StaxRip\NVEncC64_4.47_test_VS2019_CUDA10_1_Update2>NVEncC64.exe --check-device
DeviceId #0:

D:\DVD Tools\StaxRip\NVEncC64_4.47_test_VS2019_CUDA10_1_Update2>NVEncC64.exe --check-environment
NVEncC (x64) 4.47 (r1188) by rigaya, Sep  2 2019 22:03:53 (VC 1922/Win/avx2)
  [NVENC API v9.0, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
Environment Info
OS : Windows 10 x64 (18362)
CPU: Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
RAM: Used 5188 MB, Total 16341 MB
GPU: #0: [436.2]
```

> ---------------- Error Video encoding using NVEnc 4.47 ----------------
> 
> Video encoding using NVEnc 4.47 failed with exit code: 1 (0x1)
> 
> The exit code might be a system error code: STATUS_WAIT_1
> 
> The exit code might be a system error code: Incorrect function.
> 
> 
> ------------------- Video encoding using NVEnc 4.47 -------------------
> 
> "D:\DVD Tools\StaxRip\NVEncC64_4.47_test_VS2019_CUDA10_1_Update2\NVEncC64.exe" --vbrhq 3000 --codec h265 --preset quality --bframes 0 --lookahead 32 -i "D:\input.mkv" -o "D:\output.h265"
> 
> Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
> Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
> GPU #0 () does not support H.265/HEVC encoding.
> Unknown erro occurred during checking GPU.
> 
> ------------------- Video encoding using NVEnc 4.47 -------------------
> 
> "D:\DVD Tools\StaxRip\NVEncC64_4.47_test_VS2019_CUDA10_1_Update2\NVEncC64.exe" --vbrhq 3000 --codec h265 --preset quality --bframes 0 --lookahead 32 -i "D:\input.mkv" -o "D:\output.h265"
> 
> Error on nvEncOpenEncodeSessionEx: 2 (NVENC that devices pass by the client is not supported)
> Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 2 (NVENC that devices pass by the client is not supported)
> GPU #0 () does not support H.265/HEVC encoding.
> Unknown erro occurred during checking GPU.

**Comment by rigaya** [comment]
> I am not sure what else to try, as it should only be the Nvidia drivers that need to be installed, correct?

I also can't think of any more if you have tried uninstalling and then installing the driver.

**Comment by Kapiroto** [comment]
PS G:\Downloads\NVEncC_4.59_Win32> .\NVEncC.exe --check-hw
.
.
.
"This error might occur when shortage of memory, or when trying to encode more than 2 streams by NVEnc.
In Geforce, simultaneous encoding is limited up to 2, due to the NVIDIA's driver limitation.
Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 10 (NVENC indicates that the API call failed because it was unable to allocate enough memory to perform the requested operation.)"
No NVEnc support.

What it implied from this error message was that NVENC is limited to being used by a single program only.
I did a following test, Geforce Expirience uses NVENC to be able to do the recordings for example. I terminated the NVidia Share processes and rerun NVEncC.exe --check-hw and to my surprise ...

nvEncodeAPICreateInstance (APIVer = 0x71020009): Success.
NvEncOpenEncodeSessionEx (device_type = NV_ENC_DEVICE_TYPE_CUDA): Success.
nvEncDestroyEncoder: success.
Closing perf monitor ...
Evaluable Codec (s)
H.264 / AVC
H.265 / HEVC

So I'm guessing if you have other programs using NVENC, it will give this error message, as the Nvidia driver is limited.
It was the logic I discovered.

**Comment by bumpaneer** [comment]
> PS G:\Downloads\NVEncC_4.59_Win32> .\NVEncC.exe --check-hw
> .
> .
> .
> "This error might occur when shortage of memory, or when trying to encode more than 2 streams by NVEnc.
> In Geforce, simultaneous encoding is limited up to 2, due to the NVIDIA's driver limitation.
> Error on NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): 10 (NVENC indicates that the API call failed because it was unable to allocate enough memory to perform the requested operation.)"
> No NVEnc support.
> 
> What it implied from this error message was that NVENC is limited to being used by a single program only.
> I did a following test, Geforce Expirience uses NVENC to be able to do the recordings for example. I terminated the NVidia Share processes and rerun NVEncC.exe --check-hw and to my surprise ...
> 
> nvEncodeAPICreateInstance (APIVer = 0x71020009): Success.
> NvEncOpenEncodeSessionEx (device_type = NV_ENC_DEVICE_TYPE_CUDA): Success.
> nvEncDestroyEncoder: success.
> Closing perf monitor ...
> Evaluable Codec (s)
> H.264 / AVC
> H.265 / HEVC
> 
> So I'm guessing if you have other programs using NVENC, it will give this error message, as the Nvidia driver is limited.
> It was the logic I discovered.

I believe this is unrelated. In my case it is a single instance of the 64-bit application that is failing. The 32-bit version works correctly.

**Comment by bumpaneer** [comment]
Not sure when it got fixed, but it looks like my issue is solved in version 4.66.

```
D:\DVD Tools\StaxRip\NVEncC_4.66_x64>NVEncC64.exe --check-hw
NVEncC (x64) 4.65 (r1391) by rigaya, Feb 29 2020 11:28:46 (VC 1916/Win/avx2)
  [NVENC API v9.1, CUDA 10.1]
 reader: raw, avi, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP8, VP9, VC-1, MPEG1, MPEG4]
InitCuda: device #0.

Checking Environment Info...
NVEncC (x64) 4.65 (r1391) by rigaya, Feb 29 2020 11:28:46 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18363)
CPU            Intel Core i7-7700K @ 4.20GHz [TB: 4.54GHz] (4C/8T)
cuInit: Success.
cuvidInit: Success.
InitCuda: Success.
cuDeviceGetCount: Success.
gpuinfo: checking for device #0.
gpuinfo:   cuDeviceGet(0): success
gpuinfo:   cuDeviceGetName(0): GeForce GTX 1070
gpuinfo:   cudaDeviceGetAttribute: CUDA 6.1
gpuinfo:   PCIBusId: 0000:02:00.0
gpuinfo:   Got GPU Info from NVML.
gpuinfo:   NV Driver version: 441870.
gpuinfo:   CUDA Driver version: 10020.
gpuinfo: using cuda schedule mode: auto.
gpuinfo: cuCtxCreate: Success.
gpuinfo: cuCtxPopCurrent: Success.
gpuinfo: cuvidCtxLockCreate: Success.
nvenc : Loaded nvEncodeAPI64.dll.
nvenc : nvEncodeAPICreateInstance(APIVer=0x71020009): Success.
nvenc : NvEncOpenEncodeSessionEx(device_type=NV_ENC_DEVICE_TYPE_CUDA): Success.
gpuinfo:   createDeviceFeatureList
InitDeviceList: Success.
#0: GeForce GTX 1070 (1920 cores, 1835 MHz)[PCIe3x16][441.87]
Avaliable Codec(s)
H.264/AVC
H.265/HEVC
gpuinfo: Closing device #0: GeForce GTX 1070...
gpuinfo: Closing Encoder...
nvenc : nvEncDestroyEncoder...
nvenc : nvEncDestroyEncoder: success.
gpuinfo: Closed Encoder.
gpuinfo: Closed cuvid Ctx Lock...
gpuinfo: Closed cuvid Ctx Lock.
gpuinfo: Closed CUDA Context...
gpuinfo: Closed CUDA Context.
gpuinfo: Closed device #0: GeForce GTX 1070.
Closed EncodeStatus.
Closing perf monitor...
Closing logger...
```
